File: 0138 Copy List with Random Pointer/0138.py
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head: return head

        copy_map = {None: None}
        
        cur = head
        while cur:
            copy_map[cur] = Node(cur.val)
            cur = cur.next

        cur = head
        while cur:
            copy_map[cur].next = copy_map[cur.next]
            copy_map[cur].random = copy_map[cur.random]
            cur = cur.next

        return copy_map[head]

File: 0138 Copy List with Random Pointer/0138.cpp
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == nullptr) return head;

        unordered_map<Node*, Node*> copyMap;
        
        Node* cur = head;

        while (cur){
            copyMap[cur] = new Node(cur->val);
            cur = cur->next;
        }

        cur = head;

        while (cur){
            copyMap[cur]->next = copyMap[cur->next];
            copyMap[cur]->random = copyMap[cur->random];
            cur = cur->next;
        }

        return copyMap[head];
    }
};
